web/api/v1/langserver.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/prometheus-community/promql-langserver/prometheus"
+	"github.com/prometheus-community/promql-langserver/rest"
+	"github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/storage"
+)
+
+func NewLangServer(externalURL string, q storage.Queryable,
+	tr func(context.Context) TargetRetriever, logger log.Logger) (http.Handler, error) {
+	// create the custom prometheus client dedicated to the langserver
+	langserverClient := newPrometheusLangServerClient(externalURL, newWebkit(q, tr))
+	// TODO use the method CreateInstHandler once it is taking the interface prometheus.Registerer in parameter
+	// instead of the struct Registry
+	return rest.CreateHandler(context.Background(), langserverClient, logger)
+}
+
+type prometheusLangServerClient struct {
+	prometheus.Client
+	externalURL string
+	webkit      *webkit
+}
+
+func newPrometheusLangServerClient(externalURL string, webkit *webkit) prometheus.Client {
+	return &prometheusLangServerClient{
+		externalURL: externalURL,
+		webkit:      webkit,
+	}
+}
+
+// Metadata returns the first occurrence of metadata about metrics currently scraped by the metric name.
+func (c *prometheusLangServerClient) Metadata(ctx context.Context, metric string) (v1.Metadata, error) {
+	metadata := c.webkit.metadata(ctx, metric, 1)
+	if len(metadata) == 0 {
+		return v1.Metadata{}, nil
+	}
+	return v1.Metadata{
+		Type: v1.MetricType(metadata[metric][0].Type),
+		Help: metadata[metric][0].Help,
+		Unit: metadata[metric][0].Unit,
+	}, nil
+}
+
+// AllMetadata returns metadata about metrics currently scraped for all existing metrics.
+func (c *prometheusLangServerClient) AllMetadata(ctx context.Context) (map[string][]v1.Metadata, error) {
+	metadata := c.webkit.metadata(ctx, "", -1)
+	result := make(map[string][]v1.Metadata)
+	for metric, infos := range metadata {
+		result[metric] = make([]v1.Metadata, 0, len(infos))
+		for i, info := range infos {
+			result[metric][i] = v1.Metadata{
+				Type: v1.MetricType(info.Type),
+				Help: info.Help,
+				Unit: info.Unit,
+			}
+		}
+	}
+	return result, nil
+}
+
+// LabelNames returns all the unique label names present in the block in sorted order.
```

Comment:
maybe instead of using min and max for this case, we could use the hardcoded time `-100*time.Hour`

Or we just put the timerange as a parameter in the method. But that would just move the hardcoded value to the caller.

Reply:
Yeah, so I guess the language server interface doesn't provide a way to pass time parameters for any metadata lookups yet?

In an ideal world this would be a setting that the frontend could dictate, depending on what kind of lookup it is doing (e.g. depending on graph range or something like that). In a slightly less ideal world, it would be a Prometheus-wide flag. In the short-term world, it could just be hardcoded here for now to the most recent couple of hours.

Reply:
Yes for the moment the interface doesn't allow that. But I'm currently changing it with your last comments. So I can change it as well. It will give much more flexibility anyway and @slrtbtfs wanted to do it if I remember well. I just forgot a bit to change it.

Reply:
I added an attribute called `lookbackInterval` that will be used to change the interval duration.

It can be changed at the runtime if needed